documentation/proposals/Proposal - Axis Input Devices.md

```diff
@@ -0,0 +1,720 @@
+/// </summary>
+public interface IDeadzoneHandler
+{
+public float GetValue(IAxisDevice device, int descriptionIndex, Deadzone deadzone);
+public bool IsInDeadzone(IAxisDevice device, int descriptionIndex, Deadzone deadzone);
+
+    public bool IsGroupInDeadzone(IAxisDevice device, int groupIndex, Deadzone deadzone);
+    
+    // Requires that the length of the deadzones span matches the number of axes in the group
+    public bool IsGroupInDeadzone(IAxisDevice device, int groupIndex, ReadOnlySpan<Deadzone> deadzones);
+}
+```
+
+# Outputs
+Many controllers have haptics, LEDs, etc, that would be nice to access in a generic, normalized way as well. The following is a first draft concept for this functionality that could accomplish this while leaving room for different sorts of outputs later on. 
+
+Feedback requested: is custom data handling necessary here? At that point it violates the "axis" concept pretty strongly and is best served by higher-level interfaces, but I wanted to make sure we at least consider these things.
+
+```csharp
+public partial interface IAxisDevice
+{
+    /// <summary>
+    /// Output definitions for this device - commonly haptics, LEDs, etc
+    /// </summary>
+    IReadOnlyList<OutputDescription> Outputs { get; }
+    
+    /// <summary>
+    /// Allows the caller to set an output value for a specific output index.
+    /// Valid for any axis not marked as <see cref="OutputAxisTrait.RawDataOnly"/>.
+    /// </summary>
+    public void SetOutput(int index, float value);
+    
+    /// <summary>
+    /// Allows the caller to set an output value(s) for a non-floating-point output.
+    /// Valid only for axes marked as <see cref="OutputAxisTrait.RawDataOnly"/>.
+    /// Requires that the caller knows the correct data type for the output, and as such the implementation must throw
+    /// an exception if the data type is not correct.
+    /// <br/><br/>
+    /// More than likely, concrete implementations or higher-level interfaces will provide a safer way to set these
+    /// sorts of values. The absence of higher-level utilities for this method should be considered a missing feature
+    /// if the implementation utilizes <see cref = "OutputAxisTrait.RawDataOnly"/>.
+    ///<br/><br/>
+    /// It is more or less expected that this will have few, if any, implementations in the early stages of this API.
+    /// Its use is reserved for advanced use-cases or higher-level implementations where the caller knows the exact
+    /// data type required the output - for example, setting a specific bit pattern for a specific LED, driving
+    /// embedded displays, built-in speakers, etc.
+    /// </summary>
+    /// <param name="index">The index of the <see cref="OutputDescription"/> in <see cref="Outputs"/></param>
+    /// <param name="values">A pointer to the value(s) to be set</param>
+    public void SetOutput<T>(int index, T* values, int count = 1) where T : unmanaged => throw new NotImplementedException();

```

Comment:
Can we change this to be ReadOnlySpan<T>, perhaps with a DIM that implicitly creates a 1-element span from a single `ref readonly` parameter?

Reply:
funny i actually changed it from a span to a pointer when writing this - happy to change it back to appease my paranoid ass

Reply:
I think if we're splitting this up into outputs vs output groups then we should likely have a single output be a single float or a single T (with the T path being try-based to enable extension detection, again we should reserve the right to define these extensions as we see fit given that they're not the happy path, and that not everything is cleanly representable as a float.